### DISCUSSION TITLE
Micropython Classes

### DISCUSSION THREAD

**Discussion by jimcraig47** [micropython/micropython#10775] [Using MicroPython]
Have watched numerous YouTubes trying to teach myself functions and classes. Trying to apply what I've learned to creating a class to run a dcmotor for a robot car but can't get past the following error or figure out what other parameter it's looking for.

TypeError: function takes 2 positional arguments but 1 were given

class motorcontrol:
    
    def __init__(self, IN1, IN2, IN3, IN4, duty_cycle):
        self.IN1 = IN1
        self.IN2 = IN2
        self.IN3 = IN3
        self.IN4 = IN4
        self.duty_cycle = duty_cycle
        
    def forward(self, speed):
        self.speed = speed
        self.IN1.duty_u16(duty_cycle)
        self.IN1.duty_u16(0)
        self.IN3.duty_u16(duty_cycle)
        self.IN4.duty_u16(0)
        
   
    def duty_cycle(self, speed):
        if self.speed <=0 or self.speed > 100:
            duty_cycle = 0
        else:
            duty_cycle = int((speed*65536)/100)
        
                   
motorcontrol.forward(100)

**Comment by dlech** [comment]
```python
# style tips: Class names are PascalCase, constants are UPPER_SNAKE_CASE,
# everything else lower_snake_case. Leading underscore means "private" within
# module or class.

_MAX_DUTY = const(65536)

class MotorControl:
    def __init__(self, in1, in2, in3, in4):
        self.in1 = in1
        self.in2 = in2
        self.in3 = in3
        self.in4 = in4
    
    def forward(self, speed):
        duty_cycle = self._speed_to_duty_cycle(speed)
        self.in1.duty_u16(duty_cycle)
        self.in2.duty_u16(0)
        self.in3.duty_u16(duty_cycle)
        self.in4.duty_u16(0)
    
    def coast(self):
        self.in1.duty_u16(0)
        self.in2.duty_u16(0)
        self.in3.duty_u16(0)
        self.in4.duty_u16(0)
    
    def _speed_to_duty_cycle(self, speed):
        if speed < 0:
            speed = 0

        if speed > 100:
            speed = 100

        return speed * _MAX_DUTY // 100

# assuming the parameters here (IN*) are defined as PWM objects elsewhere
# have to create an instance of the class in order to use it (fixes TypeError)
motor = MotorControl(IN1, IN2, IN3, IN4)

try:
    motor.forward(100)
    # need something like time.sleep(1) here so that the program does not end right away
finally:
    # ensure that motor stops even if program crashes/is interrupted
    motor.coast()
```

**Reply by jimcraig47** [reply]
Thanks for the response, but I'm getting the same error:

_MAX_DUTY = const(65536)

class MotorControl:
    def __init__(self, in1, in2, in3, in4):
        self.in1 = in1
        self.in2 = in2
        self.in3 = in3
        self.in4 = in4
    
    def forward(self, speed):
        duty_cycle = self._speed_to_duty_cycle(speed)
        self.in1.duty_u16(duty_cycle)
        self.in2.duty_u16(0)
        self.in3.duty_u16(duty_cycle)
        self.in4.duty_u16(0)
    
    def coast(self):
        self.in1.duty_u16(0)
        self.in2.duty_u16(0)
        self.in3.duty_u16(0)
        self.in4.duty_u16(0)
    
    def _speed_to_duty_cycle(self, speed):
        if speed < 0:
            speed = 0

        if speed > 100:
            speed = 100

        return speed * _MAX_DUTY // 100
    
******************************************
from motor_control import MotorControl
from machine import Pin, PWM
from utime import sleep

freq = 1000

in1 = PWM(Pin(15))
in2 = PWM(Pin(14))
in3 = PWM(Pin(13))
in4 = PWM(Pin(12))

in1.freq(freq)
in2.freq(freq)
in3.freq(freq)
in4.freq(freq)


MotorControl.forward(100) from motor_control import MotorControl
from machine import Pin, PWM
from utime import sleep

freq = 1000

in1 = PWM(Pin(15))
in2 = PWM(Pin(14))
in3 = PWM(Pin(13))
in4 = PWM(Pin(12))

in1.freq(freq)
in2.freq(freq)
in3.freq(freq)
in4.freq(freq)

MotorControl.forward(100) #Line 18
sleep(2)
MotorControl.coast()

Traceback (most recent call last):
  File "<stdin>", line 18, in <module>
TypeError: function takes 2 positional arguments but 1 were given

**Reply by dlech** [reply]
> MotorControl.forward(100) #Line 18

You have made the same mistake. You did not create an instance of the `MotorControl` object.

```python
motor = MotorControl(in1, in2, in3, in4)
motor.forward(100)
...
```

**Reply by jimcraig47** [reply]
Right. Thanks again for your help, much appreciated.

**Comment by bulletmark** [comment]
Just some minor comments:

1. OP's original if test was:

```python
if self.speed <=0 or self.speed > 100:
    duty_cycle = 0
else:
    duty_cycle = int((speed*65536)/100)
```

The Pythonic way to write this is:

```python
if 0 < self.speed <= 100:
    duty_cycle = int((speed * 65536) / 100)
else:
    duty_cycle = 0
```

2. @dlech's replaced if test should be:

```python
if speed < 0:
    speed = 0
elif speed > 100:
    speed = 100
```
